### ISSUE TITLE
Kirby as Headless CMS

### ISSUE THREAD

**Issue by bnomei** [k-next/plugins#87]

**Comment by lukasbestle** [comment]
What is your definition of "headless"? Or to rephrase the question: How do you want to consume the data in the "headless" CMS?

The reason why I'm asking is that I see more and more people build plain old websites with frontend technologies like Vue or React and slap JS `fetch()` on the back of that site. Apparently because it's cool to use frontend technologies or because this approach seems to make it easier to animate the UI etc.

In case anyone reads this and thinks "yeah, that's exactly what I am trying to do": **Please don't**. Use content representations instead and serve the same pages both as HTML pages and as structured data for your frontend.

Of course there are still applications for headless CMS and I'd like to know what they are. :)

**Comment by joernroeder** [comment]
> yeah, that's exactly what I am trying to do 

damn, this is almost what i want to do…! jokes aside, for me a headless CMS is something where:

- data is independent from content representation
- a single data source can be consumed by lots of services (js driven webapps are just one case, there are native apps, electron apps, cli bots etc.)
- the data source / cms focuses on flexibility, (re)usability and clean UI of data chunks

 – Primary the last point is what actually convinced to go with (an even unpublished beta) kirby 3 – trust me, i've played around with lots of what comes up when you search for "headless cms" jam stack" etc. 
Most headless CMS are published these days as "Caas" (Content as a Service) where you're renting the backend as long as a) the service exists or b) the product expires. For me kirby lives (apart from a few other open source projects) in a niche as a self hosted tool to create and manage content structures easily, independently and for all sizes of clients/projects affordable in the long term. 

Personally i'm using vuejs > 80% as the tool of choice to build flexible GUIs and Apps. Therefore having Vuejs in the front- and backend helps me from staying focused by not switching context all the time. Adding custom field pligins to the panel was a no brainer for me – haven't experienced that in a while. 

The main issue we all may have with all the fancy js frontend frameworks is server side rendering – which can't be hardly replaced or simulated with content representations imo – without doing things twice or loosing the benefits of webpack, code splitting, web workers, PWAs and yes, fancy animations. There is v8js, but i haven't seen more than experiments with it.

To give you an example: the use case i'm working on right now is:

- domain with kirby: panel only manages knowledge base, lots of articles, not much changes
- content changes are pushed to gitlab content repo.
- gitlab ci starts build process: generate static html, search index etc.
- gets deployed to cdn / netlify, gitlab

by having an exchange layer (api) in between data and representation i'm quite often tend to place them on different machines as well or put other tooling like build processes in between.

Let me know what you think!

**Comment by bnomei** [comment]
i agree with @joernroeder that kirbys php backend and panel make a great headless cms - storing, managing data and a solid rest api.

but i do not agree with @lukasbestle that javascript libs should not be used instead of plain html without any progressiv enhancements in place.
my 2c: the web is getting more and more applike with service workers etc. search engines will come up with new crawlers for these websites eventually - maybe accessing a structured api instead of following links. right now it depends on the usecase and target audience.
no need for seo, target audience fullfills tech requirements, component based frontend with „fancy“ animations? use vue, react, angular.
but even if thats not the case using these for serverside static rendering can still be a good idea - since not everybody knows php by heart. junior frontend dev just seem to not know bother learning php since its more backend related and not capabale of reactive components. they can get the most out of kirby that way - skipping php. 

so what is all this talk in the slack channel about? if kirby can be used as a headless cms since it has an api then we could think about an unified way to access the api in js frontend libs. the panel does that but it does not do it in a way anybody else could reuse it - no standalone npm package.
sure it might not be much code needed to get it working but having like 10 devs do it on their own without collaborating is plain stupid and certainly not as good as it could be.

**Comment by bnomei** [comment]
using the full rest api (get/post/patch/..) should only be done with frontend login imho. like the panel does.

but apart from the js part there needs to be an easy way to access the api in a read-only mode. @rbnschlz shared his solution wrapping the php api in an GET-only route with kirby->impersonate() on slack.
the only thing thats missing here are CORS header to limit access properly - i guess. see...
https://developer.mozilla.org/en-US/docs/Web/HTTP/CORS

**Comment by lukasbestle** [comment]
> i agree with @joernroeder that kirbys php backend and panel make a great headless cms - storing, managing data and a solid rest api.

I also agree. Actually the flexibility to structure content in Kirby was the main reason why we created the content representations feature. At the same time it is of course possible to consume the data via an API.

> no need for seo, target audience fullfills tech requirements, component based frontend with „fancy“ animations? use vue, react, angular.
> but even if thats not the case using these for serverside static rendering can still be a good idea - since not everybody knows php by heart. junior frontend dev just seem to not know bother learning php since its more backend related and not capabale of reactive components. they can get the most out of kirby that way - skipping php.

I do not agree, but I am aware that both opinions are very common. It's just that I personally wouldn't want to build sites that way and would much rather invest the time in a more solid and robust solution. But that's just everyone's personal decision.

> the panel does that but it does not do it in a way anybody else could reuse it - no standalone npm package.

I agree that API clients for JS and PHP (any maybe other languages) would help a lot. We won't be able to publish official implementations before the v3 release, but we are happy to support the development of such libraries in the community.

**Comment by bnomei** [comment]
we could use https://github.com/sindresorhus/ky since its lightweight and sounds a lot liky kibry. :D

**Comment by tobystokes** [comment]
For what it's worth, having recently built a kirby site where some routes are headless and some have their html heads on, I found content representations to be a brilliant way to go. For any page, a default json representation can be just 
```
$data = $page->toArray(); 
echo json_encode($data);
```
And that's an api to fetch any content.

**Comment by bnomei** [comment]
@tobystokes how did you solve getting data from file objects? they can not have content representations, right?

**Comment by tobystokes** [comment]
They can be represented from within their page representation:
```
  $data = $page->toArray();
  $files = [];
  foreach($page->files() as $file) {
    $files[] = $file->toArray();
  }
  $data['files'] = $files;
  echo json_encode($data);
```
That's a quick way to get an overview of everything - in practice I ended up doing a bit of manipulation on some fields, like structures->yaml(), and if you want to generate img->thumbs() sizes, they'd needed to be added in there too.

**Comment by bnomei** [comment]
this sounds a lot like
https://github.com/lord-executor/kirby-json-api

but i would prefer a ready to use api instead of building my own endppints.

**Comment by joernroeder** [comment]
as some of you know i'm building API driven (static) sites and apps right now and have chosen kirby because of it's flexibility for content structuring. I'm using nuxt.js for the web front end.

Let me give you some insights what i learned, why i stopped certain experiments, how i'm doing it right now and what my ideas for API driven kirby sites in the future are:

As kirbys panel is driven by vuejs too [Nuxt.js](https://nuxtjs.org) (a vuejs framework) seemed to be a good choice for building the front end – looking at, learning from and copying concepts from the panel code was my idea as the current API docs are "just" a list of endpoints to use and some things changed anyway (network tab in the dev tools is your friend here).
So i started to read the code in the [`/panel/src/api`](https://github.com/k-next/kirby/tree/master/panel/src/api) directory to get familiar with the concepts.
The client side api seemed pretty straight forward, querying some endpoints, moving data arround. What got my attention was the line `options(id, view = "view") {` in [pages.js](github.com/k-next/kirby/blob/master/panel/src/api/pages.js#L33) – "views" 🤔? So i went to the php side of the api and got exited reading the description in [`src/Api/Model.php`](https://github.com/k-next/kirby/blob/master/src/Api/Model.php):

> The API Model class can be wrapped around any kind of object. Each model defines a set of properties that are availabel in REST calls. Those properties are defined as simple Closures which are resolved on demand. This is inspired by GraphQLs architecture and makes it possible to load only the model data that is needed for the current API call.

GraphQL? Yeah 🎉! The Model is pretty abstract, basically having a definition of _field/resolvers_, _views_ and a _selection_ which collects data to return afterwards. Perfect, so i can define my own _field resolvers_ and _views_, query them via the api and feeding the data into vuex (the vuejs store) 👌?! How to define those was my next question…

Not far i discovered [/config/api/models/Page.php](https://github.com/k-next/kirby/blob/master/config/api/models/Page.php), basically an example how to build your own resolvers for data models which i could replicate ✅. How to include my own then?

⛏⛏⛏, i came across the [`protected function setModels(array $models = null)`](https://github.com/k-next/kirby/blob/master/src/Api/Api.php#L228) which was promising again but then i came across the following line and i stopped:

```php
// /config/api.php

<?php
return [
    ...
    'models'         => include __DIR__ . '/api/models.php',
    ...
];
```

It seemed as i've found "the point of control" where models – and therefore fields, resolvers etc gets injected, even if i don't know kirbys internals, i'm sure that the data of the `models` key will be used to call the `setModels` function described above. Subclassing the API? No. I dont know how kirby does the loading of its files and apart from that, i don't see any reason to subclass the api just to add some more models 😕. What did i do?

To make it short: I'm not using all of the described awesomeness and have build a _content specific_ api with templates and  content representations – basically building my own _model views_ on different endpoints. 😤

But what's next?

With great power comes great responsibility and you guys did an awesome job building the groundwork for it – used right now by the panel. Implementing the ideas from Graphql is great, the only puzzle piece that is missing right is how others can use those tools too. Imagine two – structurally little – but important changes:

1. split up a config model like `/config/api/models/Page.php` into two files and in my oppinion also languages: Use `php` for the field config / resolvers and use `yml` for the `views` description.
2. add some kind of default like `->toArray()|->toApi()` to fields and types (presets) so people who are building apis don't have to set up all basic resolvers like slug, prev, next etc themselves.

The question really is how central to the cms blueprints are and i can imagine an `api: true` key on every field in the blueprint or the api views definition below the panel blueprint or in a separate file `site/api/page.yml`

I really like kirby and its flexible `yml` based way of building fields/sections/models and dont see any reason to use content representations at all if there is a graphql inspired system behind.

I'm curious what you think! ✌️

**Comment by bnomei** [comment]
@joernroeder thanks for explaining. so what you suggest is that a way to extend the api config – model, views [and more](https://github.com/k-next/kirby/tree/master/config/api) is need? And that some of that data could be stored in yaml instead of php files.

what might be most interesting is why the current api models did not provide the data you needed? was it too much? what endpoints were missing (like collections)? why did you need different models/views?

i am new to this reactive stuff so 🐻 with me while i wrap my head our the problem. to borrow some suggestions from another headless cms: https://www.sanity.io/docs/the-schema/donts-and-dos

- Don't model your content for display
> maybe thats the problem. kirby uses Pages concept but you might just use them to store grouped data. but "Pages" might be just be naming if one thinks about it a grouping kirbys models and blueprints instead.
- Do model according to what it "means"
> that would be template names and blueprints
- Don't organize your content for delivery
> pagination could do that, all collections have them. but its not a query language like groq.
- Do organize for versatility
> i really like the references idea here

**Comment by joernroeder** [comment]
@bnomei let me answer quickly, not writing a long post again: All of the 4 points you described are right and imo kirby fits well (pages is just not the best name) and you dont have to fully use the tree structure. What i did is setting up a "collection wrapper page" in the root, and allowing editors to add subpages (items of a specific model). sometimes the collection page has some additional fields which the collection items can use afterwards. As an example i'm using a "Team" Blueprint, children will use the "team-members" blueprint. The wrapper also defines additional data the items can use. in my case positions (a structure field) which the chosen from via a select box in the members blueprint.
BTW i'm using the autoid heavily for linking those fiels together and the page tree more as a graph structure to structure the origin of data :D

By defining models without its visual representation in mind the question would be: what data should be delivered/loaded when? this is where REST and the  one endpoint per model item can reach its limits and graphql comes into play (I'm not a REST expert btw so correct me if i'm wrong). The "which data to load when" problem cant be solved on the backend level without having a) the design in mind or b) delivering data which is may not used afterwards. 
My Team Members have a "position" and image but i may not need it to show my members e.g. in the navigation menu, i also don't want to make an http call to fetch each member separately by getting /api/members/1+n. this is where _views_ come into play and a call to /api/members?view=navigation would return an array of all members, including an identifier + the data i need. The panel uses a `compact` view to load object data before you navigate to it directly (e.g. to populate the pages field) but this is very specific as well.

I think having the ability to create multiple _views_ of a model to prevent over- and underfetching is a great starting point, does not require anyone to extend the whole api itself and can be extended to other data structures (rules, collections, translations) as well.

https://www.howtographql.com/basics/1-graphql-is-the-better-rest/
https://stackoverflow.com/questions/44564905/what-is-over-fetching-or-under-fetching

if someone is interested, i could give you access to the private repo of the project…

**Comment by bnomei** [comment]
personally what i would love to do is setup my views and models in a plugin

1) more or less like the current config and have that loaded. that might result in a lot of config values.
2) but if that gets to bloated i could also think of a simple `kirby->api()->extend($myapi)` where `$myapi` is an instance of `Kirby/Api/Api` to which i added my routes/views/model before. having a single point where to extend could make it easier to setup.

**plugin index.php**
```php
return [
    'myplugin.api' => function():Kirby/Api/Api { // will be used to extend not replace current api 
        $myapi = new Kirby/Api/Api();
        $myapi->setModels([]);
        return $myapi;

        // or merging
        $myapi = kirby()->api();
        $myapi->setModels(array_merge($myapi->models(), []));
        return $myapi;
    };
]

**Comment by bnomei** [comment]
maybe we need a headless-kit? like the starter-kit just for using kirby headless? nuxt.js seems a fine choice for an example since the panel is already using vue.

my suggestion would be using the data from the starterkit and to keep it real simple. just do static rendering, no auth, no post. plain simple querying the api.
so maybe... the startkit [shop content](https://github.com/k-next/starterkit/tree/master/content/3_shop)?

**Comment by joernroeder** [comment]
@bastianallgeier i'm pretty interested to hear your ideas on this one

**Comment by studioscholz** [comment]
Regarding the initial question by @lukasbestle:

I think it would be a shame if the Kirby dev team wouldn’t atleast explore the option in providing simple access to the REST API, since almost everything seems to be in place for it already.

I think there are definitely reasons why headless systems are so prevalent in current discussions.

Seperating content and content-consumer opens up a lot of possibilities (sharing content between various apps for example). I would also argue that both the frontend and the backend are easier to update, extend and maintain. 

Things like React, Vue, Choo.js, Nuxt, Angular, etc. are pretty fun to work with and feel very modern when used. A lot of work that previously needed to be done manually is already included. For lots of my projects, the frameworks’ downsides are negligible. 

Furthermore the web as a whole is moving away from classical websites as we know them. Github stats for the above mentioned frameworks are proof of that.

Kirby has the huge advantage of being self hosted. Whoever runs a Kirby installation is owner of his content. Most headless systems on the market today are run as a service. Two years down the road the CMS of your choice could basically disappear, because it went bankrupt or something else happened to the company running it.

Since Kirby already has a REST API to power the panel, it would make a lot of sense to make that API accessible for anyone who would like to use Kirby headless. It would basically be the only self hosted headless CMS on the market. Plus it would also beat all of the currently available ones in terms of usability and customisability.

**Comment by lukasbestle** [comment]
I think we should keep the two parts of this discussion separate:

## 1. Consuming Kirby content from a frontend framework site

I still think that this is a bad idea and the downsides are not negligible at all – progressive enhancement is a number one priority for me personally and no advantage of a framework can overweigh this. A web *site* (= a content-focused web presence that should be accessible to any kind of client technology) and a web *app* (= an interactive experience that doesn't focus on content but on "doing something") are just two very different things for me. Vue, React, Angular etc. all have a place, but it's not at the content-focused sites. Or rather: I really do support using those technologies to *augment* sites, but not to replace other, more reliable technologies. Of course that's a lot more work and less "fun", but it makes the site a better, high-quality product that doesn't collapse from the tiniest blast of air.

But as I said: This is my personal opinion and as far as I can tell also the personal opinion of several other Kirby team members, but it's *not* the "official Kirby opinion". It's an opinion of the kind "this view is my own and not the one of my employer". :)

So for the following discussion, let's assume that you have a project where you have considered all available implementation options and came to the conclusion that the Vue/React/Angular approach really is the best choice for the project (meaning: let's not discuss the advantages/disadvantages here as those depend on a lot of factors Kirby can't solve).
Instead let's now think about what Kirby would need to do to support this:

A frontend app that is going to be consumed by visitors needs "anonymous" read-only API access because you wouldn't want to ask the visitors for credentials they don't have and it's also stupid to hard-code credentials into the frontend code.
The issue with anonymous API access is the default set of permissions: Even if we said "the anonymous role should only have read-access by default" that could potentially be a security issue. Maybe there are pages with confidential content or maybe you don't want anyone to know which hidden pages exist that are going to be published later.

I have an idea though: What if we implemented an `anonymous` role that has no permissions by default. Every API client that doesn't authenticate would internally be mapped to a virtual `anonymous` user with the `anonymous` role. Because it doesn't have any permissions, the default behavior would be exactly the same as at the moment. But: Devs could give that role permissions based on the specific needs of the project.

The issue is the "virtual user". I don't know how well we would be able to implement this. But before we think about implementation details: Is this something that would be useful for you?

## 2. Using the API in "classical headless setups"

With "classical headless setup" I mean sites that have their content and their page renderer separated so that the same content can also be used for other systems outside of "the web".

Such a setup is already possible in v3 with either one of these methods:

1. You can authenticate to the API and do the same things you can do from the Panel (well, because the Panel also uses the API, which is a huge step forward from v2 in my opinion).
2. Or you can use Kirby as the page renderer as well but also provide the content as structured data using content representations.

We are definitely working on making such setups easier, for example with API client implementations, simple authentication, API docs etc. We won't have those ready for v3.0, but it's definitely something we want to work on in the future.

But what I want to point out: All of that stuff isn't going to help for "frontend framework sites". It's a different use-case, which is why I have split my response into the two paragraphs so that we can develop solutions that actually solve the problems each of the use-cases have.

**Comment by studioscholz** [comment]
> ## 1. Consuming Kirby content from a frontend framework site
> I still think that this is a bad idea and the downsides are not negligible at all – progressive enhancement is a number one priority for me personally and no advantage of a framework can overweigh this. A web _site_ (= a content-focused web presence that should be accessible to any kind of client technology) and a web _app_ (= an interactive experience that doesn't focus on content but on "doing something") are just two very different things for me. Vue, React, Angular etc. all have a place, but it's not at the content-focused sites. Or rather: I really do support using those technologies to _augment_ sites, but not to replace other, more reliable technologies. Of course that's a lot more work and less "fun", but it makes the site a better, high-quality product that doesn't collapse from the tiniest blast of air.
> 

Let’s agree to disagree. For me there have definitely been projects, where the downsides of JS frameworks are negligible. Not every website needs to be accessible to any kind of client technology.
If this were the case I would still need to supported really outdated browser versions and fringe technology that is barely used by anyone. For me this is also often a question of effort vs budget. In the past I was able to built things with Vue.js that otherwise simply wouldn’t have been possible, since they would‘ve taken too long to create without the possibilities offered by a framework.

My thoughts on what makes a website a better, high quality product is probably also vastly different from yours. My point of view may very well be a product of my background as a graphic designer and I understand why someone with a proper development background might not agree with me.

> But as I said: This is my personal opinion and as far as I can tell also the personal opinion of several other Kirby team members, but it's _not_ the "official Kirby opinion". It's an opinion of the kind "this view is my own and not the one of my employer". :)
> 

I think it would be a shame if the Kirby team would not at least consider the option of supporting ideas and viewpoints that may differ with those of their members. I also think this doesn‘t make sense from a business point of view. The more people engage with Kirby the better. This helps your bottom line as well as the community as a whole. 

> So for the following discussion, let's assume that you have a project where you have considered all available implementation options and came to the conclusion that the Vue/React/Angular approach really is the best choice for the project (meaning: let's not discuss the advantages/disadvantages here as those depend on a lot of factors Kirby can't solve).
> Instead let's now think about what Kirby would need to do to support this:
> 
> A frontend app that is going to be consumed by visitors needs "anonymous" read-only API access because you wouldn't want to ask the visitors for credentials they don't have and it's also stupid to hard-code credentials into the frontend code.
>

Yes, exactly!

> The issue with anonymous API access is the default set of permissions: Even if we said "the anonymous role should only have read-access by default" that could potentially be a security issue. Maybe there are pages with confidential content or maybe you don't want anyone to know which hidden pages exist that are going to be published later.
> 

This is the status quo for a current project I am working on. Luckily there is no chance of any confidental content (it’s a portfolio website). The biggest risk would be that someone may gain access to unpublished projects, but even that could not be considered a security risk since any data would be non-critical.

Though, I agree that there are other use cases where this is a valid concern.

> I have an idea though: What if we implemented an `anonymous` role that has no permissions by default. Every API client that doesn't authenticate would internally be mapped to a virtual `anonymous` user with the `anonymous` role. Because it doesn't have any permissions, the default behavior would be exactly the same as at the moment. But: Devs could give that role permissions based on the specific needs of the project.
> 

I initially had the same idea when first looking into the possibilities of turning Kirby into a headless CMS. My first idea was to implement a user `api@domain.com` and restrict its access to certain pages only. Ultimately this fell short, since user permissions are currently not fine grained enough to really make this work.

> The issue is the "virtual user". I don't know how well we would be able to implement this. But before we think about implementation details: Is this something that would be useful for you?

Yes, that would definitely be great (although you'll probably disagree with my reasons for why I think so). Kirby 3 so far has really been amazing to use and I am glad that I can have such open discussions around critical topics such as this one with the people behind it. Without Kirby I may have never gotten into web dev in the first place.

**Comment by studioscholz** [comment]
To keep things as short and to the point as possible:

> Maybe we are mean the same thing but talk past each other. 
> 
Yes, I think so! 🙃
 
>What we are trying to do right now is to find a solution for this in Kirby, but before that we needed to understand what the actual feature requirements are. Even though I have personal opinions, the goal is to find solutions that make sense and that's definitely not something that can be based on just my personal opinions. So I fully agree that we as the Kirby team should consider other viewpoints, which is what we are trying to do as good as we can right now. :)
> 

This is exactly what I mean when I say I value the opportunity for discussion you guys provide. Definitely a sensible way of approaching all of this.

> What's most important for us when building features is that they are flexible and not focused on just one use-case. An `anonymous` role would be such a solution and would go beyond a feature that would just be called "headless mode" in a marketing document. Anonymous API access can be useful for a lot of other use-cases as well, which is awesome! I think this is part of what makes Kirby so versatile and at the same time still simple and easy to learn.
> 

Agreed!

> > Ultimately this fell short, since user permissions are currently not fine grained enough to really make this work.
> 
> That's a good point. For the `anonymous` role to work we'd probably need per-page permissions of some kind.

I think the  `anonymous` role would need:

- Restriction to GET only
- No panel access
- Some sort of system in the style of `.gitignore`. Ideally devs would be able to restrict access to `site`-, `users`-, `translation`, `system` and `page` endpoints of Kirby. Maybe even an option to exclude certain fields within a page that is otherwise accessible or an option to filter content.

No idea if this is possible and how much work it would entail though. 😶

**Comment by joernroeder** [comment]
@studioscholz @lukasbestle maybe i'm missing the discussion here as i've just read this quickly on mobile but at least from my perspective i've described an – in the background already (partly) implemented – approach how to extend the existing api via yaml files etc. 
you could easily restrict stuff there, make it read only or even hide it completely – all by using the power of the graphql paradim which kirby is already following.
personally, i really don't care what budgets other devs have, if they are using kirby to render HTMLv1, js-web-frameworks, feed an Android/iOS App or create a PDF invoice system in the end. Kirbys job stops when it sends out the response with the data requested, nowadays in JSON – everything else is personal opinion imo and the freedom of exploration that we as devs have and should use ;)
If what i've proposed is not clear yet, i'm more than happy to describe in more detail, otherwise please make an assessment how you see my _yaml blueprints for API construction_ idea from a tech point of view.
thanks!

**Comment by babyPrince** [comment]
Hey y'all, +1 on the out of the box request. Curious to see how this would turn out!

**Comment by lukasbestle** [comment]
@joernroeder I like your ideas to provide developer APIs for API extension. Making the API extendable with custom endpoints and models for plugins or even for the specific site would be very useful indeed. We will keep your ideas in mind, but it's also not something that will be ready for v3.0.

However, if I read your posts correctly, I think that your proposal doesn't solve the "anonymous API access" issue because with a setup like proposed by you, we could only *extend* the API so that the Panel still works, correct? So we would still need something for anonymous API access.

**Comment by bnomei** [comment]
@lukasbestle maybe we could simply allow anonymous access if JS and API are called from same domain/subdomain?

**Comment by rasteiner** [comment]
in most browsers (all, by now?) the same origin policy applies by default. This doesn't however prevent anyone from accessing content that might not be intended to be public. If people get access to the api, they could for example get a list of the unlisted pages. They also could DoS attack the website with seemingly innocent but difficult requests (like calling $site->index() multiple times on the *wrong* website could bring many of them down).

Maybe something is going over my head right now, I honestly didn't read all comments above, but why do we need a core supported "headless mode"? 

![tenor](https://user-images.githubusercontent.com/6684137/46467372-da549700-c7cd-11e8-97d3-f69afbd77167.gif)

I mean, going headless is basically done by declaring a single default static template for all pages, and creating routes for the resources we want to expose. Writing the logic to get to the resources in PHP or JS isn't that different: if you need to get the latest 10 blog posts you still need to write somewhere all the sorting and filtering and limiting logic (be it in PHP or JS doesn't really change the difficulty). 
So why not write it on the server where it can't be tinkered with and serve the resources in an easy ad-hoc api (like `GET /my-api/posts/latest-10`)?

**Comment by bnomei** [comment]
@rasteiner 
the topic is two folded in my opinion.
1) how to make anonymous calls to core api?
2) is core api enough or do we need a way to extend it to make kirby properly headless capable?

about 1) this a a purely technical question. i can and has be done but it was not a smooth as is could be. so feedback has been given to the dev about this.

about 2)
its not about writing the url for the endpoint but how to get the data kirby knows about the objects inside the endpoint – the views. the question is "what to return about my posts?". using plain `toArray()` is overkill and/or not what you want, `opengraph` out of the question for core.
the problem continues if you think about images, files and their properties. building custom endpoints via routing also seems a bit like a hack since tha current api implementation already solves these problems. but just for core kirby objects. there are views/models etc. plugins devs just need a way to extend the existing api.

**Comment by bnomei** [comment]
@rasteiner 
to protect the website the api belongs to from DoS attacks it needs some (configurable) limits when allowing anonymous access. but this is not something new but a general feature of headless cms.
like here: https://www.sanity.io/docs/data-store/technical-limits

**Comment by bnomei** [comment]
@rasteiner 
also a personal comment about writing custom endpoints for reactive webapps. i did that with kirby 2. a simple test and a full blown app. to be honest if i had known it was so much boring code to create the proper json responses i would not  have picked kirby and right now i would not. 😢 just try something like sanity and you will know why.

i like a lot about kirby (especially the panel and how easy it is to create plugins) but having proper headless support at some point would be great and make it my choice in almost all projects. needless to say that more projects done with kirby will result in more sold licenses.

**Comment by lukasbestle** [comment]
> maybe we could simply allow anonymous access if JS and API are called from same domain/subdomain?

That doesn't solve the issue of default permissions. Actually I think the anonymous API access should be open to *all* requests, but in any case we need permissions. I agree with @rasteiner here.

> Maybe something is going over my head right now, I honestly didn't read all comments above, but why do we need a core supported "headless mode"?

That's exactly right. We don't.

What we need (in my opinion) is proper anonymous API access, which goes much further than the "headless mode" discussion.

The main advantage of the built-in API is that it already supports many useful features and also there could be official API clients for it so that users don't need to reinvent the wheel. That's why I think it is useful to go in this direction.

---

Maybe we should close this incredibly long and opinionated discussion and create smaller, more manageable issues in the `kirby` repo. What do you think?

**Comment by studioscholz** [comment]
> Maybe we should close this incredibly long and opinionated discussion and create smaller, more manageable issues in the `kirby` repo. What do you think?

Whatever you guys (Kirby dev team) think makes the most sense!

**Comment by lukasbestle** [comment]
Well, I don't want to cut the discussion, but I think we are at a point where pretty much all arguments have been shared. If not, feel free to post further comments. I will close (and move) this issue in a few days.

**Comment by joernroeder** [comment]
Sadly the discussion went way to emotional instead of focusing on technology and already made decisions. As i said there are great concepts hidden in core api but right now they are unexplained so it’s hard to grasp why did things the way you did like why you followed e.g. graphql conventions. If community could understand your reasons we could collectively find solutions based on best practices. 

https://statamic.com for example also seems to build its core and panel via yaml files, is based on laravel and also can be used headless. But I havent looked at it in much detail...

**Comment by bnomei** [comment]
@lukasbestle could you please create a feature suggestion issue for making the api extendable by plugins (including views) and link it here?

@joernroeder i know you already did that sort of above, but could you summarize in a few bulletpoints what your ideal workflow for extending the api would be. like where to store api yaml files, extending existing views, defining endpoints. it would help a great deal imho. thanks.

**Comment by rasteiner** [comment]
I think I understand. 

Exposing the options query language to the frontend (Bastians proposed this idea once) is rather easy and also gives an opportunity to hook into each execution step, allowing us to create whatever security model we might need for each case.

The opportunity here is that "extending" the api is done by adding page methods, field methods, models (or, really, whatever), making those extensions available also for the serverside. 
Also, viceversa, any plugin that adds methods to pages automatically exposes that functionality also to the api, without even noticing.

would be powerful and doesn't need interventions from core. Other than maybe the [demote functionality](https://github.com/k-next/kirby/issues/1052) here. 

Here's a video of me playing in the console:
https://youtu.be/Lcm2INyuzNc

**Comment by lukasbestle** [comment]
Here's the list of ideas I have extracted from this discussion:

- Official API clients (https://github.com/k-next/kirby/issues/1106)
- Support for custom API models (https://github.com/k-next/kirby/issues/1107)
- Anonymous API access (https://github.com/k-next/kirby/issues/1109)
- More granular user permissions (per page, per template) for API (https://github.com/k-next/kirby/issues/420)

I'm closing this issue now. If you have further ideas, please open a new issue over at the [kirby](https://github.com/k-next/kirby) repo.
